File: app/src/main/java/com/prey/actions/triggers/TriggerUtil.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2019 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers;

import com.prey.PreyLogger;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import org.json.JSONObject;

public class TriggerUtil {

    static SimpleDateFormat sdf_dh = new SimpleDateFormat("yyyyMMddHHmmss");
    static SimpleDateFormat sdf_h = new SimpleDateFormat("HHmmss");
    static SimpleDateFormat sdf_d = new SimpleDateFormat("yyyyMMdd");

    public static boolean haveRange(List<TriggerEventDto> listEvents) {
        for (int j = 0; listEvents != null && j < listEvents.size(); j++) {
            TriggerEventDto event = listEvents.get(j);
            if (TimeTrigger.RANGE_TIME.equals(event.type)) {
                return true;
            }
            if (TimeTrigger.REPEAT_RANGE_TIME.equals(event.type)) {
                return true;
            }
        }
        return false;
    }

    public static boolean validRange(List<TriggerEventDto> listEvents) {
        for (int j = 0; listEvents != null && j < listEvents.size(); j++) {
            TriggerEventDto event = listEvents.get(j);
            if (TimeTrigger.RANGE_TIME.equals(event.type)) {
                return validRange(event);
            }
            if (TimeTrigger.REPEAT_RANGE_TIME.equals(event.type)) {
                return validRangeTime(event);
            }
        }
        return false;
    }

    public static boolean validRangeTime(TriggerEventDto event) {
        Date now = new Date();
        Calendar cal = Calendar.getInstance();
        cal.setTime(now);
        int day_od_week = cal.get(Calendar.DAY_OF_WEEK);
        PreyLogger.d("day_od_week:" + day_od_week);
        try {
            JSONObject root = new JSONObject(event.info);
            String daysOfWeek = root.getString("days_of_week");
            daysOfWeek=daysOfWeek.replace("[","");
            daysOfWeek=daysOfWeek.replace("]","");
            String[] parts = daysOfWeek.split(",");
            String hour_fromSt = root.getString("hour_from");
            String hour_untilSt = root.getString("hour_until");
            int hour_from = Integer.parseInt(hour_fromSt);
            int hour_until = Integer.parseInt(hour_untilSt);
            int until = -1;
            try {
                until = root.getInt("until");
                PreyLogger.d("until:" + until);
                String fechaSt = sdf_d.format(now);
                int fecha = Integer.parseInt(fechaSt);
                if (fecha > until) {
                    PreyLogger.d("fecha>until");
                    return false;
                } else {
                    PreyLogger.d("fecha<=until");
                }
            } catch (Exception e) {
                PreyLogger.e("Error:"+e.getMessage(),e);
            }
            boolean isDay = false;
            for (int i = 0; parts != null && i < parts.length; i++) {
                String daySt = parts[i];
                int day=TriggerUtil.dayTrigger(daySt);
                if (day == day_od_week)
                    isDay = true;
            }
            PreyLogger.d("isDay:" + isDay);
            String horaSt = sdf_h.format(now);
            int hora = Integer.parseInt(horaSt);
            PreyLogger.d("horaSt:" + horaSt);
            if (isDay) {
                boolean a = false;
                if (hour_from <= hora) {
                    PreyLogger.d("a:");
                    a = true;
                }
                boolean b = false;
                if (hour_until >= hora) {
                    PreyLogger.d("b:");
                    b = true;
                }
                if (a && b) {
                    return true;
                }
            }
        } catch (Exception e) {
            PreyLogger.e("error validRangeTime:"+e.getMessage(),e);
        }
        return false;
    }

    public static boolean validRange(TriggerEventDto event) {
        try {
            Date nowDate = new Date();
            String nowSt = sdf_d.format(nowDate);
            double now = Double.parseDouble(nowSt);
            JSONObject root = new JSONObject(event.info);
            double from = root.getDouble("from");
            double until = root.getDouble("until");
            boolean a = false;
            if (from <= now) {
                a = true;
            }
            boolean b = false;
            if (until >= now) {
                b = true;
            }
            if (a && b) {
                return true;
            }
        } catch (Exception e) {
            PreyLogger.e("error validRange:"+e.getMessage(),e);
        }
        return false;
    }

    public static boolean validateTrigger(TriggerDto trigger) {
        Date now = new Date();
        List<TriggerEventDto> listEvents = TriggerParse.TriggerEvents(trigger.getEvents());
        for (int j = 0; listEvents != null && j < listEvents.size(); j++) {
            TriggerEventDto event = listEvents.get(j);
            if (TimeTrigger.EXACT_TIME.equals(event.type)) {
                boolean valid=true;
                try {
                    JSONObject json = new JSONObject(event.info);
                    String dateTime = null;
                    dateTime = json.getString("date");
                    PreyLogger.d(String.format("TimeTrigger dateTime:%s", dateTime));
                    Date date = TimeTrigger.EXACT_TIME_FORMAT_SDF.parse(dateTime);
                    //increased to 15
                    valid = validDateAroundMinutes(date, 15);
                } catch (Exception e) {
                    PreyLogger.e(String.format("Error :%s", e.getMessage()), e);
                }
                return valid;
            }
            if (TimeTrigger.REPEAT_TIME.equals(event.type)) {
                try {
                    JSONObject json = new JSONObject(event.info);
                    String daysOfWeek = json.getString("days_of_week");
                    daysOfWeek = daysOfWeek.replace("[", "");
                    daysOfWeek = daysOfWeek.replace("]", "");
                    String[] parts = daysOfWeek.split(",");
                    Calendar cal=Calendar.getInstance();
                    cal.setTime(now);
                    int dayNow=cal.get(Calendar.DAY_OF_WEEK);
                    String hourSt = json.getString("hour");
                    String minuteSt = json.getString("minute");
                    String secondSt = json.getString("second");
                    int hour =0;
                    try{hour=Integer.parseInt(hourSt);}catch (Exception e){PreyLogger.e("Error:"+e.getMessage(),e);}
                    int minute =0;
                    try{minute=Integer.parseInt(minuteSt);}catch (Exception e){PreyLogger.e("Error:"+e.getMessage(),e);}
                    int second =0;
                    try{second=Integer.parseInt(secondSt);}catch (Exception e){PreyLogger.e("Error:"+e.getMessage(),e);}
                    Calendar calender = Calendar.getInstance();
                    calender.set(Calendar.HOUR_OF_DAY, hour);
                    calender.set(Calendar.MINUTE, minute);
                    calender.set(Calendar.SECOND, 0);
                    calender.set(Calendar.MILLISECOND, 0);
                    Date dateTime=calender.getTime();
                    PreyLogger.d("TimeTrigger dateTime:" + dateTime);
                    boolean valid=false;
                    for (int i = 0; parts != null && i < parts.length; i++) {
                        String daySt = parts[i];
                        int day = dayTrigger(daySt);
                        if(day==dayNow){
                            valid=true;
                            PreyLogger.d("TimeTrigger day==dayNow");
                        }
                    }
                    if (valid) {
                        //increased to 15
                        valid = validDateAroundMinutes(dateTime, 15);
                        PreyLogger.d("TimeTrigger validDateAroundMinutes");
                    }
                    return valid;
                } catch (Exception e) {
                    PreyLogger.e("Error:"+e.getMessage(),e);
                }
            }
        }
        return true;
    }

    public static int dayTrigger(String daySt) {
        int day=-1;
        switch(daySt)
        {
            case "0":
                day = Calendar.SUNDAY;
                break;
            case "1":
                day = Calendar.MONDAY;
                break;
            case "2":
                day = Calendar.TUESDAY;
                break;
            case "3":
                day = Calendar.WEDNESDAY;
                break;
            case "4":
                day = Calendar.THURSDAY;
                break;
            case "5":
                day = Calendar.FRIDAY;
                break;
            default:
                day = Calendar.SATURDAY;
        }
        return day;
    }


    public static boolean validDateAroundMinutes(Date date,int minutes) {
        Date now = new Date();
        boolean valid=true;
        Calendar cal = Calendar.getInstance();
        cal.setTime(now);
        cal.add(Calendar.MINUTE, -2);
        Date lessMinutes = cal.getTime();
        long datetime=date.getTime();
        if (datetime < lessMinutes.getTime()) {
            PreyLogger.d("less minutes");
            valid=false;
        }
        cal.setTime(now);
        cal.add(Calendar.MINUTE, minutes);
        Date moreMinutes = cal.getTime();
        if (datetime > moreMinutes.getTime()) {
            PreyLogger.d("more minutes");
            valid=false;
        }
        return valid;
    }

}

File: app/src/main/java/com/prey/actions/triggers/TriggerUtil.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers

import com.prey.PreyLogger
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date

object TriggerUtil {
    var sdf_dh: SimpleDateFormat = SimpleDateFormat("yyyyMMddHHmmss")
    var sdf_h: SimpleDateFormat = SimpleDateFormat("HHmmss")
    var sdf_d: SimpleDateFormat = SimpleDateFormat("yyyyMMdd")

    fun haveRange(listEvents: List<TriggerEventDto>?): Boolean {
        var j = 0
        while (listEvents != null && j < listEvents.size) {
            val event = listEvents[j]
            if (TimeTrigger.RANGE_TIME == event.getType()) {
                return true
            }
            if (TimeTrigger.REPEAT_RANGE_TIME == event.getType()) {
                return true
            }
            j++
        }
        return false
    }

    fun validRange(listEvents: List<TriggerEventDto>?): Boolean {
        var j = 0
        while (listEvents != null && j < listEvents.size) {
            val event = listEvents[j]
            if (TimeTrigger.RANGE_TIME == event.getType()) {
                return validRange(event)
            }
            if (TimeTrigger.REPEAT_RANGE_TIME == event.getType()) {
                return validRangeTime(event)
            }
            j++
        }
        return false
    }

    fun validRangeTime(event: TriggerEventDto): Boolean {
        val now = Date()
        val cal = Calendar.getInstance()
        cal.time = now
        val day_od_week = cal[Calendar.DAY_OF_WEEK]
        PreyLogger.d("day_od_week:$day_od_week")
        try {
            val root = JSONObject(event.getInfo())
            var daysOfWeek = root.getString("days_of_week")
            daysOfWeek = daysOfWeek.replace("[", "")
            daysOfWeek = daysOfWeek.replace("]", "")
            val parts =
                daysOfWeek.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val hour_fromSt = root.getString("hour_from")
            val hour_untilSt = root.getString("hour_until")
            val hour_from = hour_fromSt.toInt()
            val hour_until = hour_untilSt.toInt()
            var until = -1
            try {
                until = root.getInt("until")
                PreyLogger.d("until:$until")
                val fechaSt = sdf_d.format(now)
                val fecha = fechaSt.toInt()
                if (fecha > until) {
                    PreyLogger.d("fecha>until")
                    return false
                } else {
                    PreyLogger.d("fecha<=until")
                }
            } catch (e: Exception) {
                PreyLogger.e("Error:" + e.message, e)
            }
            var isDay = false
            var i = 0
            while (parts != null && i < parts.size) {
                val daySt = parts[i]
                val day = dayTrigger(daySt)
                if (day == day_od_week) isDay = true
                i++
            }
            PreyLogger.d("isDay:$isDay")
            val horaSt = sdf_h.format(now)
            val hora = horaSt.toInt()
            PreyLogger.d("horaSt:$horaSt")
            if (isDay) {
                var a = false
                if (hour_from <= hora) {
                    PreyLogger.d("a:")
                    a = true
                }
                var b = false
                if (hour_until >= hora) {
                    PreyLogger.d("b:")
                    b = true
                }
                if (a && b) {
                    return true
                }
            }
        } catch (e: Exception) {
            PreyLogger.e("error validRangeTime:" + e.message, e)
        }
        return false
    }

    fun validRange(event: TriggerEventDto): Boolean {
        try {
            val nowDate = Date()
            val nowSt = sdf_d.format(nowDate)
            val now = nowSt.toDouble()
            val root = JSONObject(event.getInfo())
            val from = root.getDouble("from")
            val until = root.getDouble("until")
            var a = false
            if (from <= now) {
                a = true
            }
            var b = false
            if (until >= now) {
                b = true
            }
            if (a && b) {
                return true
            }
        } catch (e: Exception) {
            PreyLogger.e("error validRange:" + e.message, e)
        }
        return false
    }

    fun validateTrigger(trigger: TriggerDto): Boolean {
        val now = Date()
        val listEvents = TriggerParse.TriggerEvents(trigger.getEvents()!!)
        var j = 0
        while (listEvents != null && j < listEvents.size) {
            val event = listEvents[j]
            if (TimeTrigger.EXACT_TIME == event.getType()) {
                var valid = true
                try {
                    val json = JSONObject(event.getInfo())
                    var dateTime: String? = null
                    dateTime = json.getString("date")
                    PreyLogger.d(String.format("TimeTrigger dateTime:%s", dateTime))
                    val date = TimeTrigger.EXACT_TIME_FORMAT_SDF.parse(dateTime)
                    //increased to 15
                    valid = validDateAroundMinutes(date, 15)
                } catch (e: Exception) {
                    PreyLogger.e(String.format("Error :%s", e.message), e)
                }
                return valid
            }
            if (TimeTrigger.REPEAT_TIME == event.getType()) {
                try {
                    val json = JSONObject(event.getInfo())
                    var daysOfWeek = json.getString("days_of_week")
                    daysOfWeek = daysOfWeek.replace("[", "")
                    daysOfWeek = daysOfWeek.replace("]", "")
                    val parts = daysOfWeek.split(",".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray()
                    val cal = Calendar.getInstance()
                    cal.time = now
                    val dayNow = cal[Calendar.DAY_OF_WEEK]
                    val hourSt = json.getString("hour")
                    val minuteSt = json.getString("minute")
                    val secondSt = json.getString("second")
                    var hour = 0
                    try {
                        hour = hourSt.toInt()
                    } catch (e: Exception) {
                        PreyLogger.e("Error:" + e.message, e)
                    }
                    var minute = 0
                    try {
                        minute = minuteSt.toInt()
                    } catch (e: Exception) {
                        PreyLogger.e("Error:" + e.message, e)
                    }
                    var second = 0
                    try {
                        second = secondSt.toInt()
                    } catch (e: Exception) {
                        PreyLogger.e("Error:" + e.message, e)
                    }
                    val calender = Calendar.getInstance()
                    calender[Calendar.HOUR_OF_DAY] = hour
                    calender[Calendar.MINUTE] = minute
                    calender[Calendar.SECOND] = 0
                    calender[Calendar.MILLISECOND] = 0
                    val dateTime = calender.time
                    PreyLogger.d("TimeTrigger dateTime:$dateTime")
                    var valid = false
                    var i = 0
                    while (parts != null && i < parts.size) {
                        val daySt = parts[i]
                        val day = dayTrigger(daySt)
                        if (day == dayNow) {
                            valid = true
                            PreyLogger.d("TimeTrigger day==dayNow")
                        }
                        i++
                    }
                    if (valid) {
                        //increased to 15
                        valid = validDateAroundMinutes(dateTime, 15)
                        PreyLogger.d("TimeTrigger validDateAroundMinutes")
                    }
                    return valid
                } catch (e: Exception) {
                    PreyLogger.e("Error:" + e.message, e)
                }
            }
            j++
        }
        return true
    }

    fun dayTrigger(daySt: String?): Int {
        var day = -1
        day = when (daySt) {
            "0" -> Calendar.SUNDAY
            "1" -> Calendar.MONDAY
            "2" -> Calendar.TUESDAY
            "3" -> Calendar.WEDNESDAY
            "4" -> Calendar.THURSDAY
            "5" -> Calendar.FRIDAY
            else -> Calendar.SATURDAY
        }
        return day
    }


    fun validDateAroundMinutes(date: Date, minutes: Int): Boolean {
        val now = Date()
        var valid = true
        val cal = Calendar.getInstance()
        cal.time = now
        cal.add(Calendar.MINUTE, -2)
        val lessMinutes = cal.time
        val datetime = date.time
        if (datetime < lessMinutes.time) {
            PreyLogger.d("less minutes")
            valid = false
        }
        cal.time = now
        cal.add(Calendar.MINUTE, minutes)
        val moreMinutes = cal.time
        if (datetime > moreMinutes.time) {
            PreyLogger.d("more minutes")
            valid = false
        }
        return valid
    }
}